guidance/library/_json.py

```diff
@@ -858,58 +946,23 @@ def json(
     else:
         raise TypeError(f"Unsupported schema type: {type(schema)}")
 
-    definitions: Mapping[str, Callable[[], GrammarFunction]] = {}
-    if isinstance(schema, Mapping):
-        for dk in DEFS_KEYS:
-            if dk in schema:
-                assert len(definitions) == 0, "Found duplicate definitions"
-                definitions = _build_definitions(schema[dk])
+    if separators is None:
+        separators = (", ", ": ")
+
+    if whitespace_flexible:
+        skip_regex = r"[\x20\x0A\x0D\x09]+"
+        # Strip whitespace from separators since we'll handle whitespace ourselves
+        separators = (separators[0].strip(), separators[1].strip())
+    else:
+        skip_regex = None
```

Comment:
Could use some input on this. Is it sensible to allow users to pass both `separators` and `whitespace_flexible`? If so, is it sensible to strip whitespace off of the separators?

Reply:
That's a bit tricky.... the best guide would probably be what `json.dumps()` does when you set both `separators` and `indent`, but I know that's not quite the same (especially for commas). I don't think we should get too stuck on that point, though, so long as we produce valid JSON. If users have a format which absolutely must be followed, then they can always do `json.dumps(json.loads(....), separators=...., indent=.....)` themselves.

Reply:
Agreed... also, planning on allowing `indent` to be passed in the future (mirroring the json.dumps interface), but we need to add a bit of machinery to the low-level parser code to support that. I don't think it's actually imperative we give the whitespace_flexible option, especially once we enable indent. But I am not super confident on that, as we really don't know how whitespace-rigidity affects distribution shift...

Reply:
For `dumps`, the behavior is that `separators` defaults to 
- `(", ", ": ")` if `indent` is `None`
- `(",", ": ")` otherwise

`indent` makes no modifications to `separators` if they are passed by the user.

It may be sensible to do something similar with `whitespace_flexible`, defaulting to `(",", ":")` but otherwise "trusting" that the user means what they say when they pass their own. I think that non-`None` values of `indent` should raise exceptions when combined with `whitespace_flexible` though. What do you think @riedgar-ms?

Reply:
That sounds like a reasonable approach. I really wouldn't get too hung up on this, though:

1. So long as we produce valid JSON, the users will have all manner of tools for reformatting it later
2. Any extra constraints on the output need to have documentation to the effect that forcing a particular whitespace format (in all forms) risks pushing the model 'off distribution' and then refer them back to the previous point